### PR TITLE
ipn/{ipnauth,ipnlocal,ipnserver,localapi}: start baby step toward moving access checks from the localapi.Handler to the LocalBackend

### PR DESCRIPTION
**Pull request by nickkhyl**
Currently, we use the `PermitRead`/`PermitWrite`/`PermitCert` permission flags to determine which operations are allowed for a `LocalAPI` client. These checks are performed when `localapi.Handler` handles a request. Additionally, certain operations (e.g., changing the serve config) requires the connected user to be a local admin. This approach is inherently racey and is subject to TOCTOU issues. We consider it to be more critical on Windows environments, which are inherently multi-user, and therefore we prevent more than one OS user from connecting and utilizing the `LocalBackend` at the same time. However, the same type of issues is also applicable to other platforms when switching between profiles that have different `OperatorUser` values in `ipn.Prefs`.

We'd like to allow more than one Windows user to connect, but limit what they can see and do based on their access rights on the device (e.g., an admin or not) and to the currently active `LoginProfile` (e.g., owner/operator or not)., while preventing TOCTOU issues on Windows and other platforms. Therefore, we'd like to pass an actor from the `LocalAPI` to the `LocalBackend` to represent the user performing the operation. The `LocalBackend`, or the `profileManager` down the line, will then check the actor's access rights to perform a given operation on the device and against the current (or the target) profile.

This PR does not change the current permission model in any way, but it introduces the concept of an actor and includes some preparatory work to pass it around. Temporarily, the `ipnauth.Actor` interface has methods like `IsLocalSystem` and `IsLocalAdmin`, which are only relevant to the current permission model. It also lacks methods that will actually be used in the new model. We'll be adding these gradually in the next PRs and removing the deprecated methods and the `Permit*` flags at the end of the transition.

Updates tailscale/corp#18342

### CODE REVIEW

File: ipn/ipnserver/actor.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	return err == nil && token.IsLocalSystem()
+}
+
+// connIsLocalAdmin reports whether the connected client has administrative
+// access to the local machine, for whatever that means with respect to the
+// current OS.
+//
+// This is useful because tailscaled itself always runs with elevated rights:
+// we want to avoid privilege escalation for certain mutative operations.
+func connIsLocalAdmin(logf logger.Logf, ci *ipnauth.ConnIdentity, operatorUID string) bool {
```
Comment (nickkhyl) [inline ipn/ipnserver/actor.go:134]: This has been moved from https://github.com/tailscale/tailscale/commit/b78df4d48ab81e211650d33fcbdaecb7b6f6b87d/ipn/localapi/localapi.go#L1075

File: ipn/ipnserver/actor.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+func (a *actor) UserID() ipn.WindowsUserID {
+	return a.ci.WindowsUserID()
+}
+
+func (a *actor) Pid() int {
+	return a.ci.Pid()
+}
+
+// Username implements [ipnauth.Actor].
+func (a *actor) Username() (string, error) {
```
Comment (nickkhyl) [inline ipn/ipnserver/actor.go:64]: And this is https://github.com/tailscale/tailscale/blob/b78df4d48ab81e211650d33fcbdaecb7b6f6b87d/ipn/localapi/localapi.go#L1139-L1165

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -5012,7 +5005,9 @@ func (b *LocalBackend) ResetForClientDisconnect() {
 	b.setNetMapLocked(nil)
 	b.pm.Reset()
 	if b.currentUser != nil {
-		b.currentUser.Close()
+		if c, ok := b.currentUser.(io.Closer); ok {
```
Comment (maisem) [inline ipn/ipnlocal/local.go:5008]: either define your own interface like `ActorCloser` or add a `Close` func to the `Actor` interface with docs that its typically a no-op but implementations can choose to define one if they need it.

File: ipn/ipnserver/actor.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+func (a *actor) IsLocalAdmin(operatorUID string) bool {
+	return a.isLocalSystem || connIsLocalAdmin(a.logf, a.ci, operatorUID)
+}
+
+// UserID implements [ipnauth.Actor].
+func (a *actor) UserID() ipn.WindowsUserID {
+	return a.ci.WindowsUserID()
+}
+
+func (a *actor) Pid() int {
```
Comment (maisem) [inline ipn/ipnserver/actor.go:59]: maybe unexport?

File: ipn/ipnserver/actor.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	actor, err := newActor(logf, c)
+	if err != nil {
+		return context.WithValue(ctx, actorContextKey{}, err)
+	}
+	return context.WithValue(ctx, actorContextKey{}, actor)
+}
+
+// actorFromContext returns an [actor] associated with ctx,
+// or an error if the context does not carry an actor's identity.
+func actorFromContext(ctx context.Context) (*actor, error) {
```
Comment (maisem) [inline ipn/ipnserver/actor.go:110]: use `ctxkey` https://pkg.go.dev/tailscale.com/util/ctxkey

**Comment by maisem** [comment]
cc @bradfitz

File: ipn/ipnserver/actor.go (hunk 5 above)
Comment (nickkhyl) [reply ipn/ipnserver/actor.go:110]: Done, thanks!

File: ipn/ipnlocal/local.go (hunk 3 above)
Comment (nickkhyl) [reply ipn/ipnlocal/local.go:5008]: I went with the `ActorCloser` interface, thanks!

**Pull request merged**